### DISCUSSION TITLE
OpenGL extra functions

### DISCUSSION THREAD

**Discussion by Notrum666** [AvaloniaUI/Avalonia#5959] [Q&A]
1) why mandatory "GenVertexArrays", "BindVertexArray" and "DeleteVertexArray" functions are not in GLInterface class? even in avalonia opengl render example they are used via some GLExtrasInterface and attributes without much explanation
2) i need "GetActiveUniform" and "DetachShader" functions but i can't add them to GLExtrasInterfacec class from example because i don't know what to write into GLExtrasEntryPoint attribute arguments, where do i search for values for "entry" and "extension" arguments for functions i need?

**Comment by kekekeks** [comment]
1) ah, yes, we should probably remove them
2) https://docs.gl

**Reply by Notrum666** [reply]
1. i meant adding them to GLInterface..
2. there is no answer on my question - where can i find values for "entry" and "extension" arguments for "GLExtrasEntryPoint" attribute? there is no such info in the link you provided, for example - i haven't found words "glGenVertexArraysOES" or "GL_OES_vertex_array_object" (it's the values passed as arguments to GLExtrasEntryPoint attribute in render example) on glGenVertexArrays page

**Reply by kekekeks** [reply]
1. GlInterface is meant for Avalonia internal purposes, not for user-facing OpenGL bindings. So there will be as less functions there as possible
2. `glGenVertexArraysOES` is from [GL_OES_vertex_array_object](https://www.khronos.org/registry/OpenGL/extensions/OES/OES_vertex_array_object.txt) extension to OpenGL ES 1.1 and 2.0, since `glGenVertexArrays` is not a part of the standard. Thus not all OpenGL ES 2.0 implementations are required to implement it, so it is available as an extension.  OpenGL ES 3.0 has native support for `glGenVertexArrays`. You can find the list of OpenGL and OpenGL ES extensions here: https://www.khronos.org/registry/OpenGL/extensions/
